### ISSUE TITLE
Request delay when search abort in costmatrix

### ISSUE THREAD

**Issue by monteirodom** [valhalla/valhalla#3610]
In costmatrix, a threshold is checked independently for backward (targets to sources) and forward (sources to targets) searches and each search is aborted independently to each other. Thus, when a (forward)search is aborted from a source, the (backward)searches emerging from the targets do continue till that very source. Yet, if there is a cost threshold from one origin to all the targets, the threshold should hold also from the targets to that origin (am I wrong?).
And actually, when this occurs (which is often the case when there is a ferry edge near the source/target because of #3605) there is a huge delay for the target to complete, but with null results for all durations/distances.

The impact of this penalty delay of the request (for no results) is huge when it happens with a single source/target.
Even if this issue arises more often because of #3605, I think that the search to a cancelled origin/target should also be cancelled.

A potential fix could be .
* For backward search :
```
        BackwardSearch(i, graphreader);
        if (target_status_[i].threshold == 0) {
+          for (uint32_t source = 0; source < source_count_; source++) {
+              auto& targets = source_status_[source].remaining_locations;
+              auto it = targets.find(i);
+              if (it != targets.end())
+                  targets.erase(it);
+              if (targets.empty() && source_status_[source].threshold > 0) {
+                  source_status_[i].threshold = -1;
+                  if (remaining_sources_ > 0) {
+                      remaining_sources_--;
+              }
+          }
+       }
        target_status_[i].threshold = -1;
        if (remaining_targets_ > 0) {
            remaining_targets_--;
        }
```

* For forward search :
```
        ForwardSearch(i, n, graphreader);
        if (source_status_[i].threshold == 0) {
+          for (uint32_t target = 0; target < target_count_; target++) {
+          auto& sources = target_status_[target].remaining_locations;
+          auto it = sources.find(i);
+          if (it != sources.end())
+              sources.erase(it);
+          if (sources.empty() && target_status_[target].threshold > 0) {
+              target_status_[i].threshold = -1;
+              if (remaining_targets_ > 0) {
+                  remaining_targets_--;
+              }
+          }
+       }
        source_status_[i].threshold = -1;
        if (remaining_sources_ > 0) {
            remaining_sources_--;
        }
```

**Comment by kevinkreiser** [comment]
@monteirodom what you're saying makes sense to me, would you like to pr this?

**Comment by monteirodom** [comment]
Closed by #3611
